### DISCUSSION TITLE
Show assets in CP from private S3 Bucket with Temporary URLs

### DISCUSSION THREAD

**Discussion by nicorttngr** [statamic/cms#5187] [Q&A] [answered]
Hi all,
i'm using temporary Urls for displaying images from our private S3 Bucket, which are transformed into a StreamDownload.

```
$path = 'asset.jpg';
$url = Storage::disk('s3)->temporaryUrl(
    $path, now()->addSeconds(30)
);

return response()->streamDownload(function () use ($url) {
    echo file_get_contents($url);
}, 'asset.jpg');
```
However in the control panel, the images when clicked on are still broken and return 404. The smallt humbs are shown, though. 
Can I hook in the CP Asset Display and change the URL to also use the temporaryUrl? Or is there any other solution?

**Comment by edalzell** [comment, marked as answer]
ya I made an addon to do this, but haven't released it yet: https://github.com/edalzell/private-eye

**Reply by nicorttngr** [reply]
Thanks! I imported it and the frontend works, but the images in the cp are still broken.. 404 if AWS_URL is "/" or 403 if AWS_URL is the bucket with endpoint URL. 
And regarding the Tag: What's the difference between the result of download and show?

**Reply by edalzell** [reply]
`download` is to download a private asset, like a PDF or something.

`show` creates a publicly accessible url you can use to access.

**Reply by nicorttngr** [reply]
Ah okay, that's really helpful for what I've trying to do, thank you!
So I guess I have to wait for correct images in CP until your addon is released? But isn't this a common problem, when the bucket is private?

**Reply by nicorttngr** [reply]
Nevermind, the problem comes with my S3 Hoster. I switched to AWS and the images in cp are working fine using your plugin. Thank you very much!
